### PR TITLE
Add zstd and support for GNUInstallDirs

### PR DESCRIPTION
**Pull request by byrnHDF**

### CODE REVIEW

File: ZSTD/example/test/test.sh [hunk 1]
```diff
@@ -0,0 +1,261 @@
+#! /bin/sh
+#
+# Copyright by The HDF Group.
+# Copyright by the Board of Trustees of the University of Illinois.
```
Comment (lrknox) [inline ZSTD/example/test/test.sh:4]: Line 4 should be removed, and paragraph below replaced with update

File: ZSTD/example/test/testCM.sh [hunk 2]
```diff
@@ -0,0 +1,261 @@
+#! /bin/sh
+#
+# Copyright by The HDF Group.
+# Copyright by the Board of Trustees of the University of Illinois.
```
Comment (lrknox) [inline ZSTD/example/test/testCM.sh:4]: Line 4 should be removed, and paragraph below replaced with update

File: config/cmake/binex/config/cmake/HDFMacros.cmake [hunk 3]
```diff
@@ -447,5 +486,36 @@ macro (HDF_DIR_PATHS package_prefix)
       set (CMAKE_RUNTIME_OUTPUT_DIRECTORY ${EXECUTABLE_OUTPUT_PATH})
     endif ()
   endif ()
+
+#-----------------------------------------------------------------------------
+# Setup pre-3.14 FetchContent
```
Comment (lrknox) [inline config/cmake/binex/config/cmake/HDFMacros.cmake:491]: CMAKE_VERSION < 3.14?

**Review by lrknox — COMMENTED** [review]
We'll need to check the Copyright headers again, a bit more closely.

**Comment by mkitti** [comment]
c-blosc.tar.gz currently contains zstd 1.4.8 within `c-blosc/internal-complibs/zstd-1.4.8`.

The current upstream of c-blosc, v1.21.1 uses zstd-1.5.0.
https://github.com/Blosc/c-blosc/tree/v1.21.1/internal-complibs/zstd-1.5.0

The version of zstd in zstd.tar.gz appears to be a clone from the dev branch. The changelog states it is a version of the unreleased zstd-1.5.3. The latest released version is zstd-1.5.2.

Perhaps we could avoid building zstd twice and use the same version for c-blosc and zstd?

silx-kit just uses the zstd-1.5.0 packaged with c-blosc:
https://github.com/silx-kit/hdf5plugin/blob/43c9acdca00aedce326d26141a960d6c73fc61b0/setup.py#L635-L640
https://github.com/silx-kit/hdf5plugin/tree/main/src/c-blosc/internal-complibs/zstd-1.5.0

**Comment by byrnHDF** [comment]
Yes that would be nice - and there should be a way to arrange it. Unfortunately, I don't have support for dependency between filters. Plus blosc has the code for sub filters built into the repo.

I will update the blosc before releases happen.

**Comment by byrnHDF** [comment]
The latest c-blosc 1.21.1 is included in the current plugins libs folder.

**Comment by mkitti** [comment]
My recommendation, for simplicity, is to use zstd 1.5.0 by copying the internal compression library out of the version bundled with blosc rather than an unreleased zstd 1.5.3 from the dev branch.

Another approach would be to upgrade the zstd used by blosc to zstd 1.5.2, the latest released version, and also use that version of the standalone zstd.

This way only a single zstd version exists within this repository. While I think the build scripts here statically link zstd into H5Zblosc and H5Zzstd, so there is no technical conflict, it would be conceptually simpler if there were a single version of zstd.

File: ZFP/docs/interfaces.rst [hunk 4]
```diff
@@ -147,11 +144,10 @@ is whether the application calls ``H5Z_zfp_initialize()`` or not.
 Fortran Interface
 -----------------
 
-A Fortran interface based on the properties interface, described above,
+Fortran equivalents for both the properties and generic interfaces, described above,
 has been added by Scot Breitenfeld of the HDF5_ group. The code that
```
[the remark this answers is not in the document]
Comment (byrnHDF) [reply ZFP/docs/interfaces.rst:148]: That is not our file

**Comment by byrnHDF** [comment]
The intention was to use the latest released version - will correct it.

**Comment by mkitti** [comment]
I proposed that c-blosc update zstd to 1.5.2 as well:
https://github.com/Blosc/c-blosc/pull/348

**Comment by mkitti** [comment]
Are we having difficulty building Zstd on Windows? I  built zstd with MSVC yesterday.

**Comment by byrnHDF** [comment]
I found this in the overnight tests:
fatal error LNK1181: cannot open input file '..\..\..\..\bin\Release\libzstd.lib'
but now I'm not sure.

**Comment by byrnHDF** [comment]
Intel build on VS2015 had this:
 Intel\_deps\plugin-build\ZSTD\ZSTD-prefix\src\ZSTD\lib\legacy\zstd_v01.c(234): error : expected a type specifier

All VS and Intel builds had the link error.
clang builds look okay.

**Comment by mkitti** [comment]
My guess here is that we might be building more of ZSTD than needed. Here is how zstd-1.5.2 it is built over in c-blosc that I just updated:

https://github.com/Blosc/c-blosc/blob/9dc93b1de7c1ff6265d0ae554bd79077840849d8/blosc/CMakeLists.txt#L119-L141

**Comment by mkitti** [comment]
Also note that c-blosc had to add `ASM` as a language due to the addition of [`huf_decompress_amd64.S`](https://github.com/facebook/zstd/blob/dev/lib/decompress/huf_decompress_amd64.S)

https://github.com/Blosc/c-blosc2/commit/a581837d9400501c09f957298e91be0eec9a0b0a#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR76

`project(blosc LANGUAGES C ASM)`

**Comment by mkitti** [comment]
Perhaps we should set `ZSTD_LEGACY_SUPPORT` to `OFF`
https://github.com/facebook/zstd/blob/db104f6e839cbef94df4df8268b5fecb58471274/build/cmake/CMakeLists.txt#L85

**Comment by byrnHDF** [comment]
I would be willing to entertain any suggestions.
I can start with ZSTD_LEGACY_SUPPORT to OFF

**Comment by byrnHDF** [comment]
fatal error LNK1181: cannot open input file '........\bin\Release\libzstd.lib'

reason is because zstd names it zstd_static.lib instead!

**Comment by byrnHDF** [comment]
Unfortunately, zstd library uses non-standard platform names for static libraries. I will need to create a special case and duplicate library handling code/macros.
STATIC libs on windows should have this name pattern "lib<name>.lib"

**Comment by mkitti** [comment]
Is that what is happening here?

https://github.com/facebook/zstd/blob/43de2aa17d5817bc566f0b8ed13c6c8283b1d8d4/build/cmake/lib/CMakeLists.txt#L113-L118

**Comment by byrnHDF** [comment]
> Is that what is happening here?
> 
> https://github.com/facebook/zstd/blob/43de2aa17d5817bc566f0b8ed13c6c8283b1d8d4/build/cmake/lib/CMakeLists.txt#L113-L118

yes, partially.  They are changing the name to zstd_static.lib, but it should go to libzstd.lib on windows and not mingw for all compilers.

**Comment by byrnHDF** [comment]
shared name would be zstd.lib for import library for zstd.dll.

**Comment by mkitti** [comment]
Here is how conda-forge builds zstandard:
https://github.com/conda-forge/zstd-feedstock/blob/main/recipe/install.bat

**Comment by mkitti** [comment]
Do we need both here? I thought we only needed the static version.

I see the current issues are arising from using Zstandard's builtin Cmake build scripts. However, looking at how [Blosc](https://github.com/Blosc/c-blosc/blob/9dc93b1de7c1ff6265d0ae554bd79077840849d8/blosc/CMakeLists.txt#L119-L141) and the [silx-kit hdf5plugins](https://github.com/silx-kit/hdf5plugin/blob/80e2f7d414ad41e60eeff3da1787528a99ccaf4f/setup.py#L665-L678) work they bypass this issue by globbing the source files together themselves rather than using the Zstandard Cmake scripts. Perhaps they came to a similar conclusion as you in the past.

**Comment by byrnHDF** [comment]
No, we only need the static - but I do have a solution that is a minimal change and allows for custom library names. I just need to figure out why the filter link is not finding the compression library, correctly on windows.

**Comment by byrnHDF** [comment]
ZSTD filter now working on Windows.

**Pull request merged**